File: scala-3/euler/src/main/scala/com/rbleggi/euler/P10.scala
package com.rbleggi.euler

// Project Euler Problem 10
// Find the sum of all the primes below two million
@main def mainP10(): Unit = {
  val limit = 2000000

  val isPrime = Array.fill(limit)(true)
  isPrime(0) = false
  isPrime(1) = false

  for (i <- 2 until math.sqrt(limit).toInt + 1) {
    if (isPrime(i)) {
      for (j <- i * 2 until limit by i) {
        isPrime(j) = false
      }
    }
  }

  val sum = isPrime.zipWithIndex.filter(_._1).map(_._2.toLong).sum
  println(s"The sum of all primes below $limit is: $sum")
}

File: kotlin/euler/src/main/kotlin/com/rbleggi/euler/P10.kt
package com.rbleggi.euler

import kotlin.math.sqrt

fun mainP10() {
    val limit = 2000000

    val isPrime = BooleanArray(limit) { true }
    isPrime[0] = false
    isPrime[1] = false

    for (i in 2..sqrt(limit.toDouble()).toInt()) {
        if (isPrime[i]) {
            for (j in i * 2 until limit step i) {
                isPrime[j] = false
            }
        }
    }

    val sum = isPrime.withIndex().filter { it.value }.sumOf { it.index.toLong() }
    println("The sum of all primes below $limit is: $sum")
}
